llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -3012,6 +3012,28 @@ struct DecomposedSelect {
 };
 } // namespace
 
+/// Folds patterns like:
+///   select c2 (select c1 a b) (select c1 b a)
+/// into:
+///   select (xor c1 c2) b a
+static Instruction *
+foldSelectOfSymmetricSelect(SelectInst &OuterSelVal,
+                            InstCombiner::BuilderTy &Builder) {
+
+  DecomposedSelect OuterSel, InnerSel;
+  if (!match(&OuterSelVal, m_Select(m_Value(OuterSel.Cond),
+                                    m_Select(m_Value(InnerSel.Cond),
+                                             m_Value(InnerSel.TrueVal),
+                                             m_Value(InnerSel.FalseVal)),
+                                    m_Select(m_Deferred(InnerSel.Cond),
+                                             m_Deferred(InnerSel.FalseVal),
+                                             m_Deferred(InnerSel.TrueVal)))))
+    return nullptr;
+
+  Value *Xor = Builder.CreateXor(InnerSel.Cond, OuterSel.Cond);
```

Comment:
It will assert when the types of `InnerSel.Cond` and `OuterSel.Cond` differ.
```
define <2 x i32> @select_of_symmetric_selects(<2 x i32> %a, <2 x i32> %b, <2 x i1> %c1, i1 %c2) {
   %sel1 = select <2 x i1> %c1, <2 x i32> %a, <2 x i32> %b
   %sel2 = select <2 x i1> %c1, <2 x i32> %b, <2 x i32> %a
   %ret = select i1 %c2, <2 x i32> %sel1, <2 x i32> %sel2
   ret <2 x i32> %ret
 }
```